libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp

```diff
@@ -122,8 +123,8 @@ void AP_InertialSensor_Backend::_notify_new_gyro_raw_sample(uint8_t instance,
         uint64_t now = AP_HAL::micros64();
         struct log_GYRO pkt = {
             LOG_PACKET_HEADER_INIT((uint8_t)(LOG_GYR1_MSG+instance)),
-            time_us   : now,
-            sample_us : sample_us?sample_us:now,
+            time_us   : sample_us?sample_us:now,
```

Comment:
I would say that there isn't any rule that defines what the timestamp is. Actually each message defines its contents and it would be possible it wouldn't even have a timestamp - I'm actually of the opinion that the dataflash library should add the timestamp for every message instead of each message doing it.

I think that usually the timestamp is of when the message is written, but there are certainly exceptions: when the EKF is doing the fast logging it uses a fixed timestamp (the time when the EKF starts doing the maths in that cycle) for all its messages so that we can then accurately replay the information the EKF had available at the time.
On the other hand, it isn't expected (and is seen as a bug) that time goes back, even between different message types.

Given all of the above I think that having both timestamps is the correct way forward.

> plus some assurance that I could detect gaps in the sample sequence

This is where I'm not sure I'm understanding what this new sample counter is for. How are you going to detect gaps? Like I said in another comment, given the place where you increment the counter, it won't be sequential, it will have gaps between them - if it is expected that the gap is constant or not I don't know.